### DISCUSSION TITLE
Problem running OpenModelica in Mac M2

### DISCUSSION THREAD

**Discussion by ejperezHI** [OpenModelica/OpenModelica#11951] [Q&A]
Hello,

I have a MacBook Air Apple M2 and I am trying to use OpenModelica (OMEdit) in this computer.

I installed XQuartz (modified the settings to enable connections), I downloaded also and ran the OpenModelica docker image as explained [here](https://openmodelica.org/download/docker/).

Also I have followed the steps explained [here](https://openmodelica.org/forum/default-topic/3490-docker-instructions-for-mac)

What is:

1. Open Terminal
2. Execute this command:
`ip=$(ifconfig en0 | grep inet | awk '$1=="inet" {print $2}')
Xhost +$ip
docker run -it -v ~/.Xauthority:/root/.Xauthority -e DISPLAY=$ip:0  openmodelica/openmodelica:v1.22.1-gui /bin/bash`

3. As it is said in the link above, check if I have access to omc. Run the command:
`omc --version`
which outputs the following:
`OpenModelica 1.22.1`
5. And then open OMEdit with this command:
`OMEdit`
Which outputs the following
`root@987df4c9e850:/# OMEdit
qt.qpa.xcb: could not connect to display 192.168.1.36:0
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, xcb.`

As you can see, it seems to detect OpenModelica, its version and all that stuff, but I cannot open and wok with OpenModelica. Any advice on how to solve this issue?

Thank you a lot in advance

**Comment by casella** [comment]
@sjoelund any hint?

**Comment by dershow** [comment]
I have a little hint, but not a solution.  The instructions [https://openmodelica.org/download/docker/](https://openmodelica.org/download/docker/) did used to work for me.  I haven't tried in a little while, and now I'm getting the same issue as above.  The only major change I can think of is I have recently upgraded to Mac OS14, and perhaps there was a permission change?  (I'm using the same docker, xquartz, docker image etc)

**Reply by ejperezHI** [reply]
Thank you for the reply!
When I click on the link you posted it shows the "404 error page not found" so I cannot see the instructions you are refering to.

However what strikes me the most is the fact that it seems to recognize openmodelica (it recognizes the "omc --version" command showing me the version of openmodelica, even shows me the commands of omc when executing "omc" command...etc) but it cannot open OpenModelica itself with "OMEdit" command, even after downloading the correct dockerfile digest from dockerhub....

Note: The dockerfile I used is [this](https://hub.docker.com/layers/openmodelica/openmodelica/v1.22.1-gui/images/sha256-6d319d98f19d42f7d952b9242c750a9fa0643ec1c784a8617192891f96adf883?context=explore&tab=vulnerabilities) by the way

**Reply by dershow** [reply]
I fixed the link (it is the same as one that you had, but I had included it just to be specific about what had worked).

The reason that the run command is somewhat messy (and that docker-om is there to try to simplify startup) is because of X11.  XQuartz runs on your Mac, not in docker.  OMEdit runs in Docker and then should connect to your X11 server (XQuartz).  For that data to flow things have to be configured correctly on both sides.  My guess is that the error is that when OMEdit starts, as it is a graphical program, it tries to connect to the X11 server to show stuff and that is what's failing.  It can't "find" the XQuartz to talk properly.  

That could be due to how things are configured in OMEdit (but this is less likely), in the docker image, in the MacOS, in XQuartz.  My guess is that it is due to an authentication change in OS14 but I'm not sure.

**Comment by dershow** [comment]
It seems that it's a known difficulty to get docker on a mac to talk to X11.  Best I can tell it's necessary to first open XQuartz and add local host access.  Just doing the following first opens XQuartz and is working for me:
`xhost + 127.0.0.1`

Also, there is a docker variable that can avoid some of the complexity of the command that you had and that is also on the [above references instruction link](https://openmodelica.org/download/docker/), which is to use as the special Docker variable host.docker.internal (there is a deprecated mac only version:  docker.for.mac.localhost) for the display.

So, this is now working for me:
` xhost + 127.0.0.1`

`docker run -it --rm -v "$HOME:$HOME" -e "HOME=$HOME" -w "$PWD" -e "DISPLAY=host.docker.internal:0" --user $UID openmodelica/openmodelica:v1.22.1-gui`

or to run the gui:

`docker run -it --rm -v "$HOME:$HOME" -e "HOME=$HOME" -w "$PWD" -e "DISPLAY=host.docker.internal:0" --user $UID openmodelica/openmodelica:v1.22.1-gui OMEdit`

And, to do your version would be:
` xhost + 127.0.0.1`

`docker run -it -v ~/.Xauthority:/root/.Xauthority -e DISPLAY=host.docker.internal:0 openmodelica/openmodelica:v1.22.1-gui /bin/bash`

I'm not sure what changed, but this is now working for me, so hopefully for you also.

**Reply by ejperezHI** [reply]
Big thanks for your reply, 
Just a pair of doubts about it.

Correct me if I am wrong:
First, I open the XQuartz terminal and execute:
`xhost + 127.0.0.1`
Then open my mac terminal and execute:
`xhost + 127.0.0.1`
And after that:
`docker run -it -v ~/.Xauthority:/root/.Xauthority -e DISPLAY=host.docker.internal:0 openmodelica/openmodelica:v1.22.1-gui /bin/bash`

Right?

Note: Why the 127.0.0.1 direction? Is this exact same value for every computer? Or do I have to check my own?

**Comment by dershow** [comment]
That's an extra step.  At least on my mac if I open a terminal and do:
`xhost + 127.0.0.1`
it takes a few seconds and opens XQuartz then it returns.  So that can be immediately followed by the docker command.

127.0.0.1 is a loopback address, so it's a computer's way of saying "me" or "self".  All (maybe some weird exceptions?) computers have that and an outside address.  So, adding it to xhost allows connections from the same computer.  
That didn't seem to used to be necessary, so it might be that xhost defaults to blocking all connections on each restart of XQuartz?

**Reply by ejperezHI** [reply]
Thanks for your reply!
I tried that and I get the same output:
`root@987df4c9e850:/# OMEdit qt.qpa.xcb: could not connect to display 192.168.1.36:0 qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found. This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem. Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, xcb.`

I think the issue is in the line that says _"Could not load the Qt platform plugin "xcb" in "" even though it was found."_

However I have tried all that in order to make this work:

- Allowed connections from networks in XQuartz
- Allow net connections in XQuartz
- Allow local connections to the visualization server X11
- Configured the DISPLAY variable of the docker container
- Verified the Docker permissions

And the output stills de same:Even if it recognizes my commands like "omc --version" and "omc", it doesn't open OMEDit.
Am I missing something?

**Comment by sjoelund** [comment]
Perhaps you need to allow the network of the docker image in xhost, then. You are not connecting to localhost after all. Or perhaps some port forwarding in docker. --network=host doesn't work on Mac, right?

**Reply by ejperezHI** [reply]
Thanks for the reply!
I just executed that command in my MAC terminal and it doesn't work

**Comment by dershow** [comment]
As I test I would try this (IP is what you sent, not a general answer):
quit XQuartz
Then, a console window do this:
`xhost + 192.168.1.36`

Then this:

`docker run -it --rm -v "$HOME:$HOME" -e "HOME=$HOME" -w "$PWD" -e "DISPLAY=host.docker.internal:0" --user $UID openmodelica/openmodelica:v1.22.1-gui OMEdit`


One other thought, if that still doesn't work.  
Go to XQuartz->Settings->Security and make sure that both are checked (Authenticate connections and Allow connections from network clients).  

Maybe that was the problem?

**Reply by ejperezHI** [reply]
Thanks for your reply!
Done the first thing and the exact same output:
`root@987df4c9e850:/# OMEdit qt.qpa.xcb: could not connect to display 192.168.1.36:0 qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found. This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem. Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, xcb.`

However when running the docker run command, as it has been happening to me, a window pops up saying that "you have opted to share your home directory with a container", a message which I click "Ok". I do not know if this is important but it seems that it is doing what it is supposed to do.

About the second thing, I already did that the first day I started trying this

**Comment by mahge** [comment]
For what is worth, in case you did not know, you can actually build and run OpenModelica on Mac natively as described [here](https://github.com/OpenModelica/OpenModelica/blob/master/README.cmake.md). That might be an alternative assuming you have no specific reason to chose docker based use.

**Comment by dershow** [comment]
That works for me.  Although, oddly, on my Mac my native build runs simulations significantly slower than the docker version.  
I assume that it's due to some compiler flags, but I've never tried to change it.

**Comment by mahge** [comment]
That is not good. First thing first, are you building a release build? By default you would get a debug build. You can change that by specifying
```
cmake  .... CMAKE_BUILD_TYPE=Release  ...
```

That said, I doubt that having a debug build would make simulations significantly slower but it is worth making sure that is not the reason before anything else.

**Reply by dershow** [reply]
I just rebuilt with Release and yes, it now runs much faster!  
Often debug doesn't do some of the optimizations that Release does, so maybe that's why?

Thanks,
